### DISCUSSION TITLE
Alpine and SortableJS touch

### DISCUSSION THREAD

**Discussion by AbdullahObaid** [alpinejs/alpine#3856] [7. Bugs (archived)] [answered]
Hello,

I encountered a strange bug while using Alpine.js with the SortableJS or Shopify Draggable plugin. While sorting an item within an x-for loop, the functionality performs as expected when a mouse is used. However, strangely, if the item is dragged via a touch input, it triggers multiple errors. It appears that Alpine attempts to render the element while it is being dragged but not able to access the current state correctly. The wired thing is that happens only when touch is used to drag items.

example: https://codepen.io/kazunari-h/pen/XWMYeLM

![image](https://github.com/alpinejs/alpine/assets/1207636/632dbc87-373c-4a73-896a-6e029dcf7c39)

**Comment by ekwoka** [comment]
always put a :key on your x-for loops.

**Reply by AbdullahObaid** [reply]
I know, the codepen I posted is not mine, I updated the codepen to add keys [here ](https://codepen.io/mobde3/pen/BaMpLQJ) , but still have the same issue on touch devices, or even in chrome dev tools when a phone or tablet is selected.

**Comment by AbdullahObaid** [comment]
If found a related issue on SortableJS https://github.com/SortableJS/Sortable/issues/2325 , I think the issue is related to a fall back element created by SortableJS on touch drag, may the solution should be at SortableJS ?

**Comment by AbdullahObaid** [comment, marked as answer]
Here's an explanation of the problem and the solution:

**Problem:**

1. When using touch drag in libraries like SortableJS, a clone of the dragged item is created (commonly called sortable-fallback). This clone is transparent and follows the finger's movement while dragging, this is different when using mouse drag where the same element is moved without creating a clone.
2. This cloned element can be created outside of the x-for loop.
3. When Alpine.js tries to render this cloned element, it may not have access to the current state, leading to issues.

**Solution:**
1. Ensure that all bound values within the cloned element default to an empty string when they are not defined. This change should be safe because it primarily affects the cloned element only which is transparent. [A better solution would be to use x-ignore ](https://github.com/alpinejs/alpine/discussions/3856#discussioncomment-7533721)
2. Utilize Sortable events instead of relying on the `@drop` event. The `@drop` event will not be fired when the cloned element is released on touch devices, it only works when the item dropped by mouse. Sorting logic and updates to the Alpine.js state can be handled within the Sortable event handlers.

**Reply by wwdes** [reply]
Hi, thank you for providing a solution. I am unable to make it work in my situation.

```
<!-- Options for Radio or Checkbox -->
<legend class="form-label">Options</legend>
<div class="sortable d-flex flex-column gap-s">
	<template x-for="(option, index) in options" :key="index">
		<div class="input-group">
			<span class="sortable-handle cursor-grab input-group-text px-1" @touchstart.prevent="console.log('test')">
				<? F::embed_icon('grip-vertical', 20) ?>
			</span>
			<input :required="input === 'radio' || input === 'checkbox'" type="text" maxlength="<? DaoPlug::$settings['fields']['option']['maxlength'] ?>" class="form-control form-control-sm" x-model="options[index]" :name="'options[' + index + ']'">
			<button type="button" class="btn d-flex align-items-center" :class="options.length > 2 ? 'btn btn-danger' : 'btn btn-secondary'" :disabled="options.length <= 2" @click="removeOption(index)"><? F::embed_icon('x-lg', 20) ?></button>
		</div>
	</template>
</div>
```

As soon as the element is dragged with touch, it throws "Alpine Expression Error: index is not defined". Ive tried some things but have had not luck in making the cloned element to default to empty string. I suspect I would need to build this for loop in some other way. Any suggestions?

**Reply by AbdullahObaid** [reply]
So, for any bounded value that has `options` or `option` or `index` as values, you should handle the case where it is not defined.
for example in your code 
`:class="options.length > 2 ? 'btn btn-danger' : 'btn btn-secondary'"` 
can be replaced by
`:class="typeof options !== 'undefined' ? (options.length > 2 ? 'btn btn-danger' : 'btn btn-secondary') : 'btn btn-secondary'"` 
You should do similar thing for any binds

**Reply by ekwoka** [reply]
Interesting. Not sure there is a way that Alpine could more easily help with this.

Is there a way you could hook into sortable to add x-ignore to the cloned element?

**Reply by wwdes** [reply]
The main problem in my code derives from  `x-model="options[index]" :name="'options[' + index + ']'"`. I did try some things but in the end was not able to make it work.  I suspect that I need to resort from using these and take another strategy, but what that strategy is I do not know.

**Reply by ekwoka** [reply]
maybe you can put x-ignore on the element before sortable clones it?

**Reply by wwdes** [reply]
`onStart ` sortable hook is the only one that lets me manipulate the temp clone in time. No more errors!

The solution:

```
onStart: function (e) {
	let fallback = document.querySelector('.input-group.sortable-fallback')
	if (fallback) {
		fallback.setAttribute('x-ignore', '')
	}
}
```

Thanks for the `x-ignore` tip!
